// @humanforce/cloud-common/dist/dynamodb/models/model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceLoglevel = exports.InstanceType = exports.InstanceStatus = void 0;
/* eslint-disable no-unused-vars */
var InstanceStatus;
(function (InstanceStatus) {
    InstanceStatus[InstanceStatus["Pending"] = 0] = "Pending";
    InstanceStatus[InstanceStatus["Approved"] = 1] = "Approved";
    InstanceStatus[InstanceStatus["Declined"] = 2] = "Declined";
})(InstanceStatus || (exports.InstanceStatus = InstanceStatus = {}));
var InstanceType;
(function (InstanceType) {
    InstanceType[InstanceType["HFCloud"] = 0] = "HFCloud";
    InstanceType[InstanceType["OnPrem"] = 1] = "OnPrem";
    InstanceType[InstanceType["Hosted"] = 2] = "Hosted";
    InstanceType[InstanceType["HFCloudWfmOff"] = 3] = "HFCloudWfmOff";
})(InstanceType || (exports.InstanceType = InstanceType = {}));
var InstanceLoglevel;
(function (InstanceLoglevel) {
    InstanceLoglevel["TRACE"] = "trace";
    InstanceLoglevel["DEBUG"] = "debug";
    InstanceLoglevel["INFO"] = "info";
    InstanceLoglevel["WARN"] = "warn";
    InstanceLoglevel["ERROR"] = "error";
    InstanceLoglevel["FATAL"] = "fatal";
})(InstanceLoglevel || (exports.InstanceLoglevel = InstanceLoglevel = {}));
