// @humanforce/cloud-common/dist/service/parameterstoreservice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParameterStoreService = void 0;
const client_ssm_1 = require("@aws-sdk/client-ssm");
class ParameterStoreService {
    constructor() {
        this._client = new client_ssm_1.SSMClient();
    }
    async getParameter(name, withDecryption = false) {
        const command = new client_ssm_1.GetParameterCommand({
            Name: name,
            WithDecryption: withDecryption
        });
        const response = await this._client.send(command);
        return response.Parameter?.Value;
    }
    async getParameters(path, withDecryption = false, recursive = false) {
        const command = new client_ssm_1.GetParametersByPathCommand({
            Path: path,
            WithDecryption: withDecryption,
            Recursive: recursive
        });
        const response = await this._client.send(command);
        const result = new Map();
        for (const p of response.Parameters) {
            if (p.Value != null) {
                result.set(p.Name, p.Value);
            }
        }
        return result;
    }
}
exports.ParameterStoreService = ParameterStoreService;
